lib/galaxy/jobs/runners/tes.py

```diff
@@ -0,0 +1,547 @@
+                }],
+            "executors": [],
+            "volumes": [mounted_dir]
+        }
+
+        return execution_script
+
+    def output_file_gen_script(self, output_files: list):
+        """
+            Generates shell script for generating output files in container
+        """
+        script = []
+
+        for file in output_files:
+            dir_name = os.path.split(file)[0]
+            script.extend(['mkdir', '-p', dir_name, '&&', 'touch', file, '&&'])
+
+        script[-1] = '\n'
+        return ' '.join(script)
+
+    def staging_out_command(self, script_path: str, output_files: list, api_url: str, work_dir: str):
+        """
+            Generates command for staging out of files
+        """
+        command = ["python", script_path, api_url, work_dir]
+
+        for file in output_files:
+            command.append(file)
+
+        return command
+
+    def env_variables(self, job_wrapper: JobWrapper):
+        """
+            Get environment variables from job_wrapper
+        """
+        env_vars = {}
+        for variable in job_wrapper.environment_variables:
+            env_vars[variable['name']] = variable['value']
+
+        for variable in job_wrapper.job_destination.env:
+            env_vars[variable['name']] = variable['value']
+
+        env_vars["_GALAXY_JOB_TMP_DIR"] = self.container_workdir
+        env_vars["_GALAXY_JOB_HOME_DIR"] = self.container_workdir
+        return env_vars
+
+    def input_url(self, api_url: str, path: str):
+        """
+            Get URL for path
+        """
+        file_link = f"{api_url}&path={path}"
+        return file_link
+
+    def input_descriptors(self, api_url: str, input_paths: list, type: str = "FILE"):
+        """
+            Get Input Descriptor for Jobfile
+        """
+        input_description = []
+
+        for path in input_paths:
+            input_description.append({
+                "url": self.input_url(api_url, path),
+                "path": path,
+                "type": type
+            })
+        return input_description
+
+    def get_job_directory_files(self, work_dir: str):
+        """
+        Get path for all the files from work directory
+        """
+        paths = []
+        for root, _, files in os.walk(work_dir):
+            for file in files:
+                paths.append(os.path.join(root, file))
+
+        return paths
+
+    def get_docker_image(self, job_wrapper: JobWrapper):
+        """
+            For getting the docker image to be used
+        """
+        remote_container = self._find_container(job_wrapper)
+        remote_image = "vipchhabra99/pulsar-lib"
+
+        if(hasattr(job_wrapper.job_destination, "params")):
+            if("default_docker_image" in job_wrapper.job_destination.params):
+                remote_image = job_wrapper.job_destination.params.get("default_docker_image")
+
+        if(hasattr(remote_container, "container_id")):
+            remote_image = remote_container.container_id
```

Comment:
So, here, I am setting the image vipchhabra99/pulsar-lib as one of the default images, firstly it will look if we have any default docker image available in the config. If it is specified, then it overwrites with the default image to be used. If the container image is specified and is not NULL, then it overwrites with the specified image. It helps to cover all possible scenarios:
If NullContainer is set for the job, and no default image is specified in the config file, then I use the vipchhabra99/pulsar-lib , reducing the complexity for TES. If the default image is specified, I use that image. Otherwise, if the image is specified for the type of image, then I am overwriting the specified image

Reply:
So it doesn't matter which container you're using ? In that case raise an exception here and mandate setting default_docker_image

Reply:
Sure I can mandate default_docker_image

Reply:
Done, Can I close this?